server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -2053,7 +2053,16 @@ public int hashCode() {
      */
     public static <T> AffixSetting<T> prefixKeySetting(String prefix, Function<String, Setting<T>> delegateFactory) {
         BiFunction<String, String, Setting<T>> delegateFactoryWithNamespace = (ns, k) -> delegateFactory.apply(k);
-        return affixKeySetting(new AffixKey(prefix), delegateFactoryWithNamespace);
+        return affixKeySetting(new AffixKey(prefix, null, null), delegateFactoryWithNamespace);
+    }
+
+    /**
+     *  Same as above but also matches the fallback prefix in addition to the prefix of the setting.
+     */
+    public static <T> AffixSetting<T> prefixKeySetting(String prefix, String fallbackPrefix, Function<String, Setting<T>> delegateFactory) {
```

Comment:
Actually just noticing, it looks like you haven't pushed `APMAgentSettings`. The new version with fallback is currently only used in tests, explaining test failures in APMAgentSettingsTests.

With respect to https://github.com/elastic/elasticsearch/pull/104324#issuecomment-1889625348 I'm wondering if it would be beneficial if this was using the "namespaced" `delegateFactory` (BiFunction). 
Though, to be honest, at the same time I'm a little surprised about the behavior of namespace. It looks like the relative key (with the prefix removed), rather than a namespace (which I would have expected to be the prefix). But In any case, that's exactly what we are struggling / failing to extract in a robust way [here](https://github.com/elastic/elasticsearch/pull/104324/files#diff-8e5ca1315006cc00c58ef42d4bfc14f50b30d11567a845f8e551ecf3e52c830bL226-L227):

```Java
(qualifiedKey) -> {
   final String[] parts = qualifiedKey.split("\\.");
   final String key = parts[parts.length - 1];
   ... 
}
```

Using the "namespaced" `delegateFactory`, this would simply be the following if i'm not mistaken.

```Java
(key, qualifiedKey) -> {
   ... 
}

Reply:
It looks like that would work.  Would you like me to make the change in this PR?

Reply:
Yes, that would be great 👍 
But just to make sure, the "namespaced" delegateFactory is working as expected? I was surprised to see it provided the relative key rather than a namespace, but I suppose my expectation is just wrong.

Reply:
It's just poorly named and documented, it working as expected.   The namespace is the key without the prefix and suffix.

Reply:
> Yes, that would be great 👍

Done.

Reply:
Thanks @stu-elastic